Add actor states definitions & transition diagram doc (#9754)
ray-project/ray: src/ray/design_docs/actor_states.rst

Actor State: Definitions & Transition Diagram
============================================

An actor can be in one of the following states:

> - **DEPENDENCIES_UNREADY**: the actor info ins registered in GCS. But its dependencies are not ready.

Reviewer: ```suggestion
- **DEPENDENCIES_UNREADY**: the actor info is registered in GCS. But its dependencies are not ready.
```

  the state will become alive again.

- **DEAD**: the actor is already dead and won't be restarted.

::

                                                                         3
   0                            1                       2          ------------->
 ---->DEPENDENCIES_UNREADY-------->PENDING_CREATION-------->ALIVE                RESTARTING
               |                            |                  |   <-------------      |
               |                            |                  |         4             |
               |                            |                  |                       |
             8 |                         7  |                6 |                       | 5
               |                            |                  |                       |
               |                            |                  |                       |
               |                            |                  |                       |
               |                            v                  |                       |
                -------------------------->DEAD<---------------------------------------

> - **0**: GCS will create an actor whose state is `DEPENDENCIES_UNREADY` when a `RegisterActor` request is received from CoreWorker.

Reviewer: ```suggestion
- **0**:  When GCS receives a `RegisterActor` request from core worker, GCS will persist the actor in database with state `DEPENDENCIES_UNREADY`.
```

Reviewer: Nit: 
1) I think it's clearer to state what event will trigger the state transition first.
2) No need to use camel case for a term (e.g., `CoreWorker`), unless that is something you can search in code (e.g, `RegisterActor`).

Author: fixed

> - **1**: GCS will change actor state to `PENDING_CREATION` when a `CreateActor` request is received from CoreWorker.

Reviewer: ```suggestion
- **1**: When core worker has finished resolving the actor dependencies, it will send a `CreateActor` request to GCS and GCS will update actor state to `PENDING_CREATION`.
```

- **2**: GCS will change actor state to `ALIVE` when a successful `PushNormalTask` reply is received from CoreWorker.

> - **3**: GCS will change actor state to `RESTARTING` when the owner worker/node of actor is dead.

Reviewer: This is triggered when the actor itself dies, not the owner dies.

- **4**: GCS will change actor state to `ALIVE` when the actor reconstruction is successful.

> - **5**: GCS will change actor state to `DEAD` when the actor remaining restarts number is 0.

Reviewer: No 5 and 7 won't happen, right?

Author: It will happen and i have updated the descriptions of 5 and 7.

Actor State: Definitions & Transition Diagram
============================================

An actor can be in one of the following states:

> - **DEPENDENCIES_UNREADY**: the actor info is registered in GCS. But its dependencies are not ready.

Reviewer: ```suggestion
- **DEPENDENCIES_UNREADY**: The actor info is registered in GCS. But its dependencies are not ready.
```

               |                            |                  |                       |
             8 |                         7  |                6 |                       | 5
               |                            |                  |                       |
               |                            |                  |                       |
               |                            |                  |                       |
               |                            v                  |                       |
                -------------------------->DEAD<---------------------------------------

- **0**: When GCS receives a `RegisterActor` request from core worker, GCS will persist the actor in database with state `DEPENDENCIES_UNREADY`.

- **1**: When core worker has finished resolving the actor dependencies, it will send a `CreateActor` request to GCS and GCS will update actor state to `PENDING_CREATION` in memory.

- **2**: When core worker has finished actor creation task, it will send a `PushTask` reply to GCS and GCS will persist the actor in database with state `ALIVE`.

> - **3**: When GCS detected that the worker/node of actor is dead and the actor remaining restarts number is greater than 0, it will persist the actor in database with state `RESTARTING`.

Reviewer: ```suggestion
- **3**: When GCS detectes that the worker/node of an actor is dead and the actor's remaining restarts number is greater than 0, it will persist the actor in database with state `RESTARTING`.
```

- **4**: When the actor is successfully reconstructed, GCS will persist the actor in database with state `ALIVE`.

- **5**: If the actor is restarting, GCS detects that its worker or node is dead and the actor remaining restarts number is 0, it will persist the actor in database with state `DEAD`.

>   If the actor is not detached, when GCS detected that the owner worker/node/actor is dead, it will persist the actor in database with state `DEAD`.

Reviewer: ```suggestion
  If an actor is not detached, when GCS detects that its owner is dead, it will update its state to `DEAD` in the database.
```

Reviewer: Can you also add a index number for the 2 cases: 
```
5: 1) ...
    2) ...

- **6**: When GCS detected that the actor is dead and the actor remaining restarts number is 0, it will persist the actor in database with state `DEAD`.
  If the actor is not detached, when GCS detected that the owner worker/node/actor is dead, it will persist the actor in database with state `DEAD`.

- **7**: If the actor is not detached, when GCS detected that the owner worker/node/actor is dead, it will persist the actor in database with state `DEAD`.

> - **8**: When GCS detected that the actor creator is dead, it will persist the actor in database with state `DEAD`.

Reviewer: ```suggestion
- **8**:  For both detached and non-detached actors, when GCS detects that an actor's creator is dead, it will update its state to `DEAD`. Because in this case, the actor can never be created.
```

- **1**: When core worker has finished resolving the actor dependencies, it will send a `CreateActor` request to GCS and GCS will update actor state to `PENDING_CREATION` in memory.

- **2**: When core worker has finished actor creation task, it will send a `PushTask` reply to GCS and GCS will update actor state to `ALIVE` in database.

- **3**: When GCS detects that the worker/node of an actor is dead and the actor's remaining restarts number is greater than 0, it will update actor state to `RESTARTING` in database.

- **4**: When the actor is successfully reconstructed, GCS will update its state to `ALIVE` in the database.

- **5**: 

  1) If the actor is restarting, GCS detects that its worker or node is dead and its remaining restarts number is 0, it will update its state to `DEAD` in database.
  
  2) If an actor is non-detached, when GCS detects that its owner is dead, it will update its state to `DEAD` in the database.

- **6**: 

  1) When GCS detected that an actor is dead and its remaining restarts number is 0, it will update its state to `DEAD` in database.
         

>   2) If the actor is non-detached, when GCS detects that its owner is dead, it will update its state to `DEAD` in the database.

Reviewer 2: When non-deatched actors are dead, and remaining start number is originally 0, it is also marked as DEAD. https://github.com/ray-project/ray/blob/2babad99066882617a564562f907df8890eaf444/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L841 (It is marked as DEAD "again" when owner is dead).

Author: line 50 describes this logic?

- **7**: If the actor is non-detached, when GCS detects that its owner is dead, it will update its state to `DEAD` in the database.

- **8**: 
  

>   1) For both detached and non-detached actors, when GCS detects that an actor's creator is dead, it will update its state to `DEAD`. Because in this case, the actor can never be created.

Reviewer 2: Is it true? Do creator / owner have different meaning in your description? I think detached actors shouldn't be killed when creator is dead.

Author: If the creator is dead when actor state is `DEPENDENCIES_UNREADY`, the dependencies will can not be resolved, so we update the actor state `DEAD`.
